lib/srv/db/postgres/redshift-activate-user.sql

```diff
@@ -0,0 +1,40 @@
+CREATE OR REPLACE PROCEDURE teleport_activate_user(username varchar, roles text)
+LANGUAGE plpgsql
+AS $$
+DECLARE
+    roles_length integer;
+    cur_roles_length integer;
+BEGIN
+    roles_length := JSON_ARRAY_LENGTH(roles);
+
+    -- If the user already exists and was provisioned by Teleport, reactivate
+    -- it, otherwise provision a new one.
+    IF EXISTS (SELECT user_id FROM svv_user_grants WHERE user_name = username AND admin_option = false AND role_name = 'teleport-auto-user') THEN
+        -- If the user has active connections, make sure the provided roles
+        -- match what the user currently has.
+        IF EXISTS (SELECT user_name FROM stv_sessions WHERE user_name = CONCAT('IAM:', username)) THEN
+          SELECT INTO cur_roles_length COUNT(role_name) FROM svv_user_grants WHERE user_name = username AND admin_option=false AND role_name != 'teleport-auto-user';
+          IF roles_length != cur_roles_length THEN
+            RAISE EXCEPTION 'TP002: User has active connections and roles have changed';
+          END IF;
+          FOR i IN 0..roles_length-1 LOOP
+            IF NOT EXISTS (SELECT role_name FROM svv_user_grants WHERE user_name = username AND admin_option=false AND role_name = JSON_EXTRACT_ARRAY_ELEMENT_TEXT(roles,i)) THEN
+                RAISE EXCEPTION 'TP002: User has active connections and roles have changed';
+            END IF;
+          END LOOP;
+          RETURN;
+        END IF;
+        -- Otherwise reactivate the user, but first strip it of all roles to
+        -- account for scenarios with left-over roles if database agent crashed
+        -- and failed to cleanup upon session termination.
+        CALL teleport_deactivate_user(username);
```

Comment:
I got following error from `teleport_activate_user` procedure call in fresh redshift setup during testing:
```
tsh db connect --db-user=marek --db-name=dev marek-redshift-cluster-1
psql: error: connection to server at "localhost" (::1), port 54170 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 54170 failed: your Teleport role requires automatic database user provisioning but an attempt to activate database user "marek" failed due to the following error: ERROR: user "marek" does not exist (SQLSTATE 42704)
```

Also when I tries to call teleport_activate_user manually: 
![Screenshot 2023-10-20 at 13 14 26](https://github.com/gravitational/teleport/assets/22402974/08060a1e-605e-4943-a886-bce9d6d6cc05)

```
dev=# select * from pg_user;
    usename     | usesysid | usecreatedb | usesuper | usecatupd |  passwd  | valuntil | useconfig
----------------+----------+-------------+----------+-----------+----------+----------+-----------
 rdsdb          |        1 | t           | t        | t         | ******** | infinity |
 admin          |      100 | t           | t        | f         | ******** |          |
 teleport-admin |      101 | f           | f        | f         | ******** |          |
```

```
dev=# select * from SVV_ROLES;
 role_id |     role_name      |   role_owner   | external_id
---------+--------------------+----------------+-------------
  105587 | sys:operator       | rdsdb          |
  105588 | sys:monitor        | rdsdb          |
  105589 | sys:dba            | rdsdb          |
  105590 | sys:secadmin       | rdsdb          |
  105591 | sys:superuser      | rdsdb          |
  106245 | role1              | admin          |
  106246 | role2              | admin          |
  106247 | role3              | admin          |
  106248 | teleport-auto-user | teleport-admin |
  ```

@greedy52 I'm missing something ?

Reply:
`GRANT ROLE "sys:superuser" TO "teleport-admin"`?

But I am surprised that it didn't fail earlier. Let me look into that

Reply:
Seems caused by missing spaces. fixed in [dc1cdc4](https://github.com/gravitational/teleport/pull/33307/commits/dc1cdc4382482fbd29171ce2828e5a22679bad6f).

Could you give it a try?

This is so weird that it worked for my user `STeve`. I think because `QUOTE_IDENT` has no effect on username with allow cases. Even then, we are supposed to see `CREATE USER` failed first before the grant. I will play around a little bit more but just Redshift weirdness.

Reply:
Yes, After [dc1cdc4](https://github.com/gravitational/teleport/pull/33307/commits/dc1cdc4382482fbd29171ce2828e5a22679bad6f) my setup works. Thanks for the alignment.